### PR TITLE
Python3 port of the C-Extension

### PR DESCRIPTION
**Pull request by stephan-hof**
As requested in #7 I open a pull request to port the C-Extension of ExtensionClass.

### CODE REVIEW

**Comment by tseaver** [comment]
@stephan-hof Wow, thank you!  I will try to get a review done this coming week (unless somebody else beats me to it).

**Comment by tseaver** [comment]
I haven't done a detailed review, but I did check that your branch shows 100% test coverage after merging with the current master (we're only measuring coverage of Python code, of course).

**Comment by stephan-hof** [comment]
I found a regression in ExtensionClass (python version and also my C version) while working on the C-Extension of Acquisition 

The old C code of ExtensionClass changed an attribute error into the following:
```
class Dummy(Base):
    pass

Dummy().a
AttributeError: a
```

The python version (+ my ported C version) does the following:
```
class Dummy(Base):
    pass

Dummy().a
AttributeError: 'Dummy' object has no attribute 'a'
```

See, the error messages contains the typename of an object.
From my point of view a port should not change behaviour, so I would change the C and python version to hide the typename again.

The python version of Acquisition struggled also with the changed behaviour of ExtensionClass.
A lot of doctests there have asserts for `AttributeError: <name>` (no typename in the error message), but instead of questioning ExtensionClass a fix to Acquisition  was introduced.
https://github.com/zopefoundation/Acquisition/blob/master/src/Acquisition/__init__.py#L745
When ExtensionClass behaves the old way Acquisition could be changed to rely on ExtensionClass to throw the appropriate AttributeError. This would make a port of the C-Extension of Acquisition easier too.


Alternative:
* Keep the regression => Keep the new style Attribute error
* Change all the doctests in Acquisition to assert the new style Attribute error
* Change python version of Acquisition to keep the new style Attribute error

@jamadden any comments on this? Since you wrote the python version of Acquisition.

**Comment by jamadden** [comment]
I don't particularly care. The only reason the code in Acquisition you pointed to is there is because of doctests; I'm not particularly a fan of doctests or relying on exact exception messages in general, but I didn't want to break anything or create too large of a PR.

Personally, I'd rather see the more informative error messages raised everywhere (although in Acquisition's case, that could be slightly misleading).

**Comment by tseaver** [comment]
We have no guarantee that Python itself doesn't break reliance on exact exception message tests:  feel free to break doctests which rely on them if if makes the code clearer / faster / more maintainable (but open issues against them, if you can).

File: tox.ini [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py27,py27-pure,py33,py34,py35,pypy,pypy3,coverage
+    py27,py27-pure,py33,py34,py34-pure,py35,pypy,pypy3,coverage
```
Comment (icemac) [inline tox.ini]: Is there a reason to test this specific Python 3 version as pure python?

Comment (stephan-hof) [inline tox.ini]: There is no specific reason why I have chosen this version. My goal was that at least one python3 version tests the python only implementation.
I think for completeness I would make sense to add a also a `py33-pure` and `py35-pure` possibly 3.6. I'll update the pull request in the next days.

Background:
Per default a test uses the C-Extension. In a `pure` environment the C-Extension is not compiled nor considered for the unittests. So only the `pure` environments test the python only implementation of ExtensionClass. Which means with the current envlist the python only implementation is tested in python27 and python34.

**Comment by hannosch** [comment]
I've reviewed this to the best of my ability and found no issues.

I've squashed and merged this manually in f6ef36a78d36180876d46941948e0d054b376ccb. In subsequent commits I've extended the tox test list of cover all combinations of pure/non-pure environments and all Python versions addressing icemac's comment.

**Comment by tseaver** [comment]
@hannosch Thanks for picking up my slack!

@stephan-hof Thanks again for your patch!

**Comment by hannosch** [comment]
@tseaver you are welcome :) @stephan-hof if you want to tackle that Acquisition C port now ... ;)

**Comment by stephan-hof** [comment]
@hannosch many thanks for reviewing it.

**Pull request closed without merging**